xarray/core/resample.py

```diff
@@ -0,0 +1,339 @@
+
+        Apply uses heuristics (like `pandas.GroupBy.apply`) to figure out how
+        to stack together the datasets. The rule is:
+        1. If the dimension along which the group coordinate is defined is
+           still in the first grouped item after applying `func`, then stack
+           over this dimension.
+        2. Otherwise, stack over the new dimension given by name of this
+           grouping (the argument to the `groupby` function).
+
+        Parameters
+        ----------
+        func : function
+            Callable to apply to each sub-dataset.
+        **kwargs
+            Used to call `func(ds, **kwargs)` for each sub-dataset `ar`.
+
+        Returns
+        -------
+        applied : Dataset or DataArray
+            The result of splitting, applying and combining this dataset.
+        """
+        kwargs.pop('shortcut', None)  # ignore shortcut if set (for now)
+        applied = (func(ds, **kwargs) for ds in self._iter_grouped())
+        combined = self._combine(applied)
+
+        return combined.rename({self._resample_dim: self._dim})
+
+    def reduce(self, func, dim=None, keep_attrs=False, **kwargs):
+        """Reduce the items in this group by applying `func` along the
+        pre-defined resampling dimension.
+
+        Note that `dim` is by default here and ignored if passed by the user;
+        this ensures compatibility with the existing reduce interface.
+
+        Parameters
+        ----------
+        func : function
+            Function which can be called in the form
+            `func(x, axis=axis, **kwargs)` to return the result of collapsing an
+            np.ndarray over an integer valued axis.
+        keep_attrs : bool, optional
+            If True, the datasets's attributes (`attrs`) will be copied from
+            the original object to the new one.  If False (default), the new
+            object will be returned without attributes.
+        **kwargs : dict
+            Additional keyword arguments passed on to `func`.
+
+        Returns
+        -------
+        reduced : Array
+            Array with summarized data and the indicated dimension(s)
+            removed.
+        """
+        return super(DatasetResample, self).reduce(
+            func, self._dim, keep_attrs, **kwargs)
+
+    def _interpolate(self, kind='linear'):
+        from .dataarray import DataArray
+
+        new_x = self._full_index.values.astype('float')
+
+        arrs = []
+        # Apply the interpolation to each DataArray in our original Dataset
+        for v in self._obj.data_vars:
+            da = self._obj[v].copy()
+
+            x = da[self._dim].astype('float')
+            y = da.values
+            axis = da.get_axis_num(self._dim)
+
+            f = interp1d(x, y, kind=kind, axis=axis, bounds_error=True,
+                         assume_sorted=True)
+
+            # drop any existing non-dimension coordinates along the resampling
+            # dimension
+            coords = OrderedDict()
```

Comment:
This works, but given that `coords` are shared between all data variables on a Dataset, it would be a little cleaner to only do this once.

Possibly we can leave this one to a follow-up if someone is annoyed by slow performance here. Can you add a comment noting that the lower-level Variable API could be used to speed things up here?

Reply:
Re-factored to do once prior to looping over all our variables to interpolate them. Looks like it works just fine, although if there's a cleaner way to get a "test" DataArray from a Dataset, that might make things a little bit prettier